drivers/net/wireless/marvell/mwifiex/pcie_quirks.c

```diff
@@ -0,0 +1,248 @@
+}
+
+static void mwifiex_pcie_set_power_d3cold(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+
+	mwifiex_dbg(adapter, MSG, "putting into D3cold...\n");
+
+	pci_save_state(pdev);
+	if (pci_is_enabled(pdev))
+		pci_disable_device(pdev);
+	pci_set_power_state(pdev, PCI_D3cold);
+}
+
+static int mwifiex_pcie_set_power_d0(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	int ret;
+
+	mwifiex_dbg(adapter, MSG, "putting into D0...\n");
+
+	pci_set_power_state(pdev, PCI_D0);
+	ret = pci_enable_device(pdev);
+	if (ret) {
+		dev_err(&pdev->dev, "pci_enable_device failed\n");
+		return ret;
+	}
+	pci_restore_state(pdev);
+
+	return 0;
+}
+
+int mwifiex_pcie_reset_d3cold_quirk(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	struct pci_dev *parent_pdev = pci_upstream_bridge(pdev);
+
+	/* Power-cycle (put into D3cold then D0) */
+	mwifiex_dbg(adapter, MSG,
+		    "Using reset_d3cold quirk to perform FW reset\n");
+
+	/* We need to perform power-cycle also for bridge of wifi because
+	 * on some devices (e.g. Surface Book 1), the OS for some reasons
+	 * can't know the real power state of the bridge.
+	 * When tried to power-cycle only wifi, the reset failed with the
+	 * following dmesg log:
+	 * "Cannot transition to power state D0 for parent in D3hot".
+	 */
+	mwifiex_pcie_set_power_d3cold(pdev);
+	mwifiex_pcie_set_power_d3cold(parent_pdev);
+
+	mwifiex_pcie_set_power_d0(parent_pdev);
+	mwifiex_pcie_set_power_d0(pdev);
+
+	return 0;
+}
+
+int mwifiex_pcie_reset_wsid_quirk(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	char *path = ACPI_WSID_PATH;
+	static guid_t wsid_dsm_guid;
+	acpi_handle handle;
+	union acpi_object *obj;
+	acpi_status status;
+
+	mwifiex_dbg(adapter, MSG,
+		    "Using reset_wsid quirk to perform FW reset\n");
+
+	guid_parse(WSID_DSM_UUID, &wsid_dsm_guid);
+
+	status = acpi_get_handle(NULL, path, &handle);
+	if (ACPI_FAILURE(status)) {
+		dev_err(&pdev->dev, "No ACPI handle for path %s\n", path);
+		return -ENODEV;
+	}
+
+	if (!acpi_has_method(handle, "_DSM")) {
+		dev_err(&pdev->dev, "_DSM method not found\n");
+		return -ENODEV;
+	}
+
+	if (!acpi_check_dsm(handle, &wsid_dsm_guid,
+			    WSID_REV, WSID_FUNC_WIFI_PWR_OFF)) {
+		dev_err(&pdev->dev,
+			"_DSM method does not support requested function\n");
+		return -ENODEV;
+	}
+
+	/* card will be removed immediately after this call on Surface 3 */
+	mwifiex_dbg(adapter, MSG, "turning wifi off...\n");
+	obj = acpi_evaluate_dsm(handle, &wsid_dsm_guid,
+				WSID_REV, WSID_FUNC_WIFI_PWR_OFF,
+				NULL);
+	if (!obj) {
+		dev_err(&pdev->dev,
+			"device _DSM execution failed for turning wifi off\n");
+		return -ENODEV;
```

Comment:
Not sure if `-ENODEV` is the best error to return here. Maybe `-EFAULT` is better, but I'm not really sure. Same with the evaluation below.

Reply:
Hmm, upstream uses various return values for `acpi_evaluate_dsm()` failure:
- return -ENXIO;
- return -ENODEV;
- return -ENOTSUPP;
- return -EINVAL;
- return -EIO;

In this case, it just means that execution has failed. So, as you mention, `-EFAULT` may be better. I'll change this in the next rebase.

By the way, it seems that I've forgotten using `ACPI_FREE(obj);`, I'll add in the next rebase.

Reply:
- changed `acpi_evaluate_dsm()` failure return value to `-EFAULT`
- added `ACPI_FREE(obj);` after `acpi_evaluate_dsm()` success

Addressed by the new rebase.

Reply:
I think `-EIO` would be fine too. `-ENODEV` and `-ENXIO` (to me) seem more like "the device or expected method is not available", which you already checked previously. `-ENOTSUPP` is similar, just indicates that that might be something that's not expected to be there. `-EINVAL` is more for input dependent errors, i.e. invalid input to the function calling the _DSM.

In our case, I think it makes more sense to use an error value that signals "the function is there, we tried to execute it, but something during execution failed".

Reply:
- https://github.com/torvalds/linux/blob/02de58b24d2e1b2cf947d57205bd2221d897193c/include/uapi/asm-generic/errno-base.h#L9
```c
#define	EIO		 5	/* I/O error */
```

- https://github.com/torvalds/linux/blob/02de58b24d2e1b2cf947d57205bd2221d897193c/include/uapi/asm-generic/errno-base.h#L18
```c
#define	EFAULT		14	/* Bad address */
```

Honestly, I'm not sure which is more appropriate in this case. So, I'll again change the return value to `-EIO` this time considering that I see many usage in upsteam.

Reply:
changed return value of `acpi_evaluate_dsm()` failure to `-EIO`